### PR TITLE
Allow the display of messages without failing

### LINKED ISSUE #28: Is there a way to not fail a test on console.error, but still print?
I would still like to see the `console.error` message when it happens, but be able to have the test *not* fail.

### PR DESCRIPTION
**Pull request by vernak2539**
This implements functionality where a message can be used to determine if the usage of `console[methodName]` should cause an error. Fixes #28.

I'm not 100% confident on the naming, happy to take steer on this. 

The call stack is a bit weird when echoed out (see below). Is this something we'd want to fix, or something that already occurs?

```
$ jest tests/index.test.js
  console.log
      console.error
        my error message that I do not control

          126 |       console[methodName] = (...args) => {
          127 |         newMethod(...args)
        > 128 |         originalMethod(...args)
              |         ^
          129 |       }
          130 |       unexpectedConsoleCallStacks.length = 0
          131 |     })

          at console.error (../../../index.js:128:9)
          at Object.<anonymous>.module.exports (index.js:2:11)
          at Object.<anonymous> (index.test.js:5:30)

      at Object.<anonymous> (tests/index.test.js:83:13)
```

I've not updated the README yet, but will do it once the direction is correct.

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -44,6 +45,8 @@ declare namespace init {
      * skip console checks from this package or not.
      */
     skipTest?: (args: { testName: string; testPath: string }) => boolean
+
+    displayMessageAndNotFail?: (message: string, methodName: ConsoleMethodName) => boolean
```
Comment (ValentinH) [inline index.d.ts:49]: I suggest naming this `allowMessage`.

File: index.js [hunk 2]
```diff
@@ -115,7 +122,11 @@ const init = ({
       shouldSkipTest = canSkipTest()
       if (shouldSkipTest) return
 
-      console[methodName] = newMethod // eslint-disable-line no-console
+      // eslint-disable-next-line no-console
+      console[methodName] = (...args) => {
+        newMethod(...args)
```
Comment (ValentinH) [inline index.js:127]: Do you want to make this change the default behavior or only if the new option returns true?

I'm open to both but just want to ensure that the implementation matches the original intent.

**Comment by ValentinH** [comment]
Regarding the stack trace, this is unfortunate indeed. It would be nice to find a solution because the point of this feature was to keep the original output of console methods 🙈

**Comment by renatodeleao** [comment]
+1 for this feature. 
Currently using `silenceMessage` as workaround to **prevent** some tests fails, but would prefer explicit pass a list of errors that **make** tests fail.

**Comment by vernak2539** [comment]
I'm really sorry about dropping the ball here. I've had some things going on and wasn't able to try to finish this off.  I'll update the naming soon 🤞 

>Regarding the stack trace, this is unfortunate indeed. It would be nice to find a solution because the point of this feature was to keep the original output of console methods 🙈

Yeah, I'm thinking to change it from how it is now, to checking for the "allowMessage" closer to where the method name is overwritten. Hopefully, this will keep the stack trace how we'd want it. This will diverge from the original implementation discussion, but I'm not sure how we could keep the stack trace with that implementation anyhow. 

If someone gets to it first, please go ahead!

**Comment by ValentinH** [comment]
I just reviewed this PR and I adjusted the code a bit because I don't think it was matching the original intent. Indeed, when using `allowConsole`, the test was not failing but the logged message wasn't the original one but the one from the lib:
image

Let me know if this is still matching your needs and then we can merge this 🙂

**Comment by vernak2539** [comment]
@ValentinH This looks much better than my version! Thanks for taking the time to update it.

I think it will fit the needs still (as the test passes)!

**Comment by ValentinH** [comment]
Released as https://github.com/ValentinH/jest-fail-on-console/releases/tag/v3.2.0

**Pull request merged**